### ISSUE TITLE
No longer able to retrieve depth texture from buffer

### ISSUE THREAD

**Issue by Epihaius** [panda3d/panda3d#212]
Some months ago I made [this code snippet](https://www.panda3d.org/forums/viewtopic.php?f=8&t=19585&sid=a753b7b7e04729a60931c9a04ae34f5d) to use the depth buffer instead of a CollisionRay to get the coordinates of a point on a model under the mouse cursor.
With more recent versions of Panda it doesn't work anymore, as it seems that the depth texture assigned to the buffer gets converted to a color texture.

Here is some code to clarify the problem:

```
from panda3d.core import *
from direct.showbase.ShowBase import ShowBase

loadPrcFileData("", "depth-bits 24")


class MyApp(ShowBase):

    def __init__(self):

        ShowBase.__init__(self)

        depth_tex = Texture("depth_texture")
        depth_tex.set_format(Texture.F_depth_component)
        fbp = FrameBufferProperties()
        fbp.set_depth_bits(24)
        buffer = self.win.make_texture_buffer("buffer", 1, 1, depth_tex, to_ram=True, fbp=fbp)
        self.graphicsEngine.render_frame()
        tex_formats = {Texture.F_rgb: "F_rgb", Texture.F_depth_component: "F_depth_component"}
        print "Texture format:", tex_formats[depth_tex.get_format()]


app = MyApp()
app.run()
```

The output is:

`Texture format: F_rgb`

With older Panda versions, it is as expected:

`Texture format: F_depth_component`

Perhaps I should use a different way to access the depth buffer texture? I already tried calling `GraphicsOutput.add_render_texture(depth_tex, GraphicsOutput.RTM_copy_ram, DrawableRegion.RTP_depth)` instead of `make_texture_buffer`, and `Texture.set_match_framebuffer_format(True)`, as well as `FrameBufferProperties.setup_depth_texture(depth_tex)`, but the results are always the same. Replacing `F_depth_component` with `F_depth_component24` doesn't help either.

This happens using version 9bfc425 of Panda3D (Python 2.7, 64-bit).

**Comment by rdb** [comment]
Thanks, this is fixed now, I believe.  The format-choosing code had been changed to set the format to `F_depth_component24`, but the copy-to-RAM code in the GL code didn't understand what component type to use for `F_depth_component24`.

**Comment by Epihaius** [comment]
Thank you!
The code snippet works again now, but I have to set the number of depth bits to 16; when it is set to 24, the depth values are all zero (when checking the pixel color of the PNMImage I store the depth texture into).

Consider the following code:

```
from panda3d.core import *
from direct.showbase.ShowBase import ShowBase

loadPrcFileData("", "depth-bits 24")


class MyApp(ShowBase):

    def __init__(self):

        ShowBase.__init__(self)

        depth_tex = Texture("depth_texture")
        depth_tex.set_format(Texture.F_depth_component)
        fbp = FrameBufferProperties()
        fbp.set_depth_bits(24)
        buffer = self.win.make_texture_buffer("buffer", 1, 1, depth_tex, to_ram=True, fbp=fbp)
        self.graphicsEngine.render_frame()
        img = PNMImage(1, 1)
        depth_tex.store(img)
        pixel = img.get_xel(0, 0)
        print("Pixel:", pixel)


app = MyApp()
app.run()
```


The output is:

`Pixel: LVecBase3f(0, 0, 0)`

but when I do `fbp.set_depth_bits(16)` I get the expected result:

`Pixel: LVecBase3f(0, 0, 1)`

**Comment by rdb** [comment]
That's because PNMImage only supports up to 16 bits per component, and the current code downloads 24-bit-depth buffers into the 32-bit `T_unsigned_int` component type.

There's no good reason why `Texture.store` doesn't work for `T_unsigned_int` textures.  I will check in a fix for that shortly.  However, do you remember what component_type the data was for your code in 1.9?

**Comment by Epihaius** [comment]
> There's no good reason why `Texture.store` doesn't work for `T_unsigned_int` textures. I will check in a fix for that shortly.

Much appreciated!

> However, do you remember what component_type the data was for your code in 1.9?

Yes, that was `T_float`.

On a related note, it seems that the `TexturePeeker` class isn't compatible with depth textures; would it be worth it to fix this (if that's even possible), in case it gave a better performance when continually checking pixel values, compared to storing the texture in a PNMImage every frame?
If you find it worthwile, I would be willing to file a new issue report for this.

P.S. (Off-topic) In case you missed it, I commented on issue 189 some time ago.

**Comment by rdb** [comment]
OK, now 24-bit depth is converted to `T_float` again (e4e24ee).  Also, `TexturePeeker` now supports both unsigned integer and floating-point textures (348b1c3).  And store() is now implemented for `T_unsigned_int` (9dcfcbf).

Indeed, `TexturePeeker` is a faster way to access texture data.  In fact, if you try to store a floating-point texture into a `PNMImage`, it will first store it into a `PfmFile` (the floating-point equivalent for PNMImage) and convert that into a `PNMImage`, so it is extra redundant.  `TexturePeeker` is even faster because it does no extra copy or conversion at all.

I did not forget about #189, I will look into that when I find some time.
